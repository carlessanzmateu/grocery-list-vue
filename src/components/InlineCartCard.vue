<template>
  <div class="inline-cart-card">
      <img
        class="image"
        :src="item.getImage()"
      />
      <div class="card-information">
        <h3 class="title">
          <strong>{{ item.getName() }}</strong>
        </h3>
        <div class="buy-info">
          <a class="add-item" @click="addItem">+</a>
          <span class="products-left">{{ item.getSelectedStock() }}</span>
          <a class="remove-item" @click="removeItem">-</a>
        </div>
      </div>
      <div class="amount">{{ item.getItemTotalPrice() }}$</div>
    </div>
</template>

<script>
export default {
  name: 'InlineCartCard',
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  methods: {
    addItem () {
      this.$emit('add-item', this.item)
    },
    removeItem () {
      this.$emit('remove-item', this.item)
    }
  }
}
</script>

<style lang="scss" scoped>
.inline-cart-card {
  display: grid;
  grid-template-columns: 75px 1fr 50px;
  width: 100%;
  height: 75px;

  .image {
    width: 75px;
    height: 75px;
  }

  .card-information {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding-left: 15px;

    .title {
      font-size: 16px;
    }

    .buy-info {
      .add-item,
      .remove-item {
        cursor: pointer;
        font-size: 20px;
      }

      .add-item {
        margin-right: 10px;
      }
      .remove-item {
        margin-left: 10px;
      }
    }
  }
  .amount {
    display: flex;
    justify-content: center;
    align-items: center;
  }
}
</style>
