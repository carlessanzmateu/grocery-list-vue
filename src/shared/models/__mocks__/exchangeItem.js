const mockExchangeItem = {
  id: '1',
  // eslint-disable-next-line
  image_url: 'mock.png',
  stock: 1,
  productName: 'some name',
  price: 10,
  productDescription: 'Some description',
  favorite: 0
}
