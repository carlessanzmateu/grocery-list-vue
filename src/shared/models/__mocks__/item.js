class MockItem {
  constructor() {
    this.$id = '1'
    this.$image = 'mock.png'
    this.$stockReadyToBuy = 0
    this.$stock = 1
    this.$name = 'some name'
    this.$price = 10
    this.$description = 'Some description'
    this.$favorite = false
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  getStock () {
    return this.$stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}

export { MockItem }
