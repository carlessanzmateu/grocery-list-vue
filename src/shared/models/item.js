export default class Item {
  $id
  $image
  $stockReadyToBuy
  $stock
  $name
  $price
  $description
  $favorite

  constructor (rawItem) {
    this.$id = rawItem.id
    this.$image = rawItem.image_url
    this.$stock = rawItem.stock
    this.$name = rawItem.productName
    this.$price = rawItem.price
    this.$description = rawItem.productDescription
    this.buildFavorite(rawItem.favorite)
    this.initSelectedStock()
  }

  getId () {
    return this.$id
  }

  getImage () {
    return this.$image
  }

  initSelectedStock () {
    this.$stockReadyToBuy = 0
  }

  getSelectedStock () {
    return this.$stockReadyToBuy
  }

  // setSelectedStock (selectedStock) {
  //   this.$stockReadyToBuy = selectedStock
  // }

  getStock () {
    return this.$stock
  }

  setStock (stock) {
    this.$stock = stock
  }

  getStockLeft () {
    return this.getStock() - this.getSelectedStock()
  }

  canIncreaseSelectedStock () {
    return this.getSelectedStock() < this.getStock()
  }

  canDecreaseSelectedStock () {
    return this.getSelectedStock() > 0
  }

  increaseSelectedStock () {
    if (!this.canIncreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy++
  }

  decreaseSelectedStock () {
    if (!this.canDecreaseSelectedStock()) {
      return
    }

    this.$stockReadyToBuy--
  }

  getName () {
    return this.$name
  }

  getPrice () {
    return this.$price
  }

  getItemTotalPrice () {
    return this.getSelectedStock() * this.getPrice()
  }

  getDescription () {
    return this.$description
  }

  buildFavorite (favorite) {
    this.$favorite = (favorite === '1')
  }

  getIsFavorite () {
    return this.$favorite
  }

  setAsFavorite () {
    this.$favorite = true
  }

  removeFromFavorite () {
    this.$favorite = false
  }

  toggleFavorite () {
    this.$favorite = !this.$favorite
  }
}
