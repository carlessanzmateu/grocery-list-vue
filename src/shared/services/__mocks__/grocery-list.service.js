export default {
  getAllGroceries: jest.fn(),
  getAllFavorites: jest.fn(),
  updateGrocery: jest.fn()
}
