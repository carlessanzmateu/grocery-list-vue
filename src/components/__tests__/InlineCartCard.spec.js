import { shallowMount } from '@vue/test-utils'
import InlineCartCard from '@/components/InlineCartCard.vue'
import { MockItem } from '@/shared/models/__mocks__/item'

describe('InlineCartCard', () => {
  let wrapper
  function createWrapper() {
    return shallowMount(InlineCartCard, {
      propsData: {
        item: new MockItem()
      }
    })
  }

  it('should create', () => {
    wrapper = createWrapper()
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should match html', () => {
    wrapper = createWrapper()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have addItem method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.addItem).toBeDefined()
  })

  it('should emit add-item event when call addItem method', () => {
    wrapper = createWrapper()

    wrapper.vm.addItem()

    expect(wrapper.emitted()['add-item'])
  })

  it('should have removeItem method', () => {
    wrapper = createWrapper()

    expect(wrapper.vm.removeItem).toBeDefined()
  })

  it('should emit remove-item event when call removeItem method', () => {
    wrapper = createWrapper()

    wrapper.vm.removeItem()

    expect(wrapper.emitted()['remove-item'])
  })
})
