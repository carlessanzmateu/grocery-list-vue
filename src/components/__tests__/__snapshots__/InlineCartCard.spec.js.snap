// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineCartCard should match html 1`] = `
<div
  class="inline-cart-card"
>
  <img
    class="image"
    src="mock.png"
  />
   
  <div
    class="card-information"
  >
    <h3
      class="title"
    >
      <strong>
        some name
      </strong>
    </h3>
     
    <div
      class="buy-info"
    >
      <a
        class="add-item"
      >
        +
      </a>
       
      <span
        class="products-left"
      >
        0
      </span>
       
      <a
        class="remove-item"
      >
        -
      </a>
    </div>
  </div>
   
  <div
    class="amount"
  >
    0$
  </div>
</div>
`;
